rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/domain/service/LRAService.java

```diff
@@ -414,60 +410,40 @@ public int renewTimeLimit(URI lraId, Long timelimit) {
         return lra.setTimeLimit(timelimit);
     }
 
-    public boolean isLocal(URI lraId) {
-        return hasTransaction(lraId);
-    }
-
     public List<LRAData> getFailedLRAs() {
         Map<URI, LongRunningAction> failedLRAs = new ConcurrentHashMap<>();
 
-        lraRecoveryModule.getFailedLRAs(failedLRAs);
+        getRM().getFailedLRAs(failedLRAs);
 
-        return failedLRAs.values().stream().map(t -> t.getLRAData()).collect(toList());
+        return failedLRAs.values().stream().map(LongRunningAction::getLRAData).collect(toList());
     }
 
-    /**
-     * When the deployment is loaded register for recovery
-     */
-    @PostConstruct
-    void enableRecovery() {
-        assert lraRecoveryModule == null;
-
-        if (LRALogger.logger.isDebugEnabled()) {
-            LRALogger.logger.debugf("LRAServicve.enableRecovery");
-        }
-
-        lraRecoveryModule = new LRARecoveryModule(this);
-        RecoveryManager.manager().addModule(lraRecoveryModule);
-        Implementations.install();
-
-        lraRecoveryModule.getRecoveringLRAs(recoveringLRAs);
-
-        for (LongRunningAction transaction : recoveringLRAs.values()) {
-            transaction.getRecoveryCoordinatorUrls(participants);
+    private LRARecoveryModule getRM() {
+        // since this method is reentrant we do not need any synchronization
+        if (recoveryModule != null) {
+            return recoveryModule;
         }
-    }
 
-    /**
-     * When the deployment is unloaded unregister for recovery
-     */
-    @PreDestroy
-    void disableRecovery() {
-        if (lraRecoveryModule != null) {
+        List<RecoveryModule> modules = recoveryPropertyManager.getRecoveryEnvironmentBean().getRecoveryModules();
 
-            RecoveryManager.manager().removeModule(lraRecoveryModule, false);
-            Implementations.uninstall();
+        for (RecoveryModule rm : modules) {
+            if (rm instanceof LRARecoveryModule) {
+                recoveryModule = (LRARecoveryModule) rm;
+                recoveryModule.getRecoveringLRAs(recoveringLRAs);
 
-            lraRecoveryModule = null;
+                for (LongRunningAction transaction : recoveringLRAs.values()) {
+                    transaction.getRecoveryCoordinatorUrls(participants);
+                }
```

Comment:
What. Why aren't you reading the code to find out the answer yourself, you could ask the same type of question for any line of code.
These two statements are initialising recoveringLRAs and participants which should be obvious. Are you asking me to redesign the code?

Reply:
@ochaloup will you check that you are okay with my responses (apart from the change that will ensure that access to `LRARecoveryModule._transactionUidVector` is thread safe) before I push a new commit?

Reply:
@mmusgrov I'm sorry as my question was probably ambiguous. I marked the exact lines of code that I don't understand about. There is probably a reason just I'm not getting it. Thus I'm asking.

My point is when the `getRM` is called then there is considered if the `recoveryModule` is null. If it's null then the module is (or may be) initialized. And that `recoveryModule` is returned as the result of the method as with the initialization the `static` variable is defined.
Now I don't understand what's the reason to call `recoveryModule.getRecoveringLRAs` and iterates over the recovering lras at time the `recoveryModule` is defined, ie. at time when `getRM is called for the first time and the recovery module was not yet initialized.

I wonder what is purpose of method call `recoveryModule.getRecoveringLRAs` and iterating over the results.

Reply:
@mmusgrov I'm ok, sure, feel free to push the new commit.

Reply:
@ochaloup  I am loosing track of which review comments are still unresolved. As far a I am concerned everything should have been addressed. Can you go through your comments and mark the ones that are resolved so that I know where I am with this epic review.

Reply:
All is resolved now. I'm just curious about these lines. But it's a question not a review abstain comment.

Reply:
Okay I will merge this one now and then we can revisit how we do the initialisation in another PR later on ...